### PR TITLE
Export: allow prepending notes with the Shaarli instance's URL

### PR DESCRIPTION
**Pull request by virtualtam**
Relates to #102

Additions:
- application:
  - export: allow prepending note permalinks with the instance's URL
  - test coverage

Modifications:
- export template: switch to an HTML form
  - link selection (all/private/public)
  - prepend note permalinks with the instance's URL

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -1600,10 +1601,17 @@ function renderPage()
 
         // export as bookmarks_(all|private|public)_YYYYmmdd_HHMMSS.html
         $selection = $_GET['selection'];
+        $prependNoteUrl = $_GET['prepend_note_url'];
```
Comment (ArthurHoaro) [inline index.php]: This probably generates a warning if the checkbox isn't check (unknown index).

**Comment by ArthurHoaro** [comment]
Neat! This fixes the oldest Shaarli's known bug!

File: index.php (hunk 1 above)
Comment (virtualtam) [inline index.php]: Thanks, fixed!

**Comment by nodiscc** [comment]
Thanks a lot!  

Very concise and understandable patch (for me). I would like to know if you did write the tests before or after writing the code?

**Comment by virtualtam** [comment]
@nodiscc I usually proceed as follows:
- define what needs to be achieved (e.g. feature to add or issue to fix)
- look for related code sections and annotate them with `TODO`s & `FIXME`s to get a quick overview of what needs to be modified
- when possible, start writing test cases that match the expected behaviour / highlight issues
- start tampering with the actual code

Writing tests before modifying code (aka TDD - Test Driven Development) is usually easy when the features are simple and you have a good knowledge of the codebase (which happened to be the case here). Also, there already were some tests for regarding data preparation for export, so adding new ones was quite straightforward ;-)

**Comment by ArthurHoaro** [comment]
Personal input: I've tried rigorous TDD, but unless you know exactly what your implementation will look like, it felt like I spent/wasted too much time rewriting tests. So usually, I:

  * write code blindly,
  * write my tests,
  * make my code work using the tests,
  * polish the whole thing and use coverage to make sure I didn't forget important parts.

**Comment by virtualtam** [comment]
Yep, TDD is rather a pain to apply on a codebase you're not familiar with, or has bad design, or has scarce test coverage. I mostly use it on software I use/develop on a daily basis, as the codebase is usually quite sane and testing-friendly ;-)

For more insight, here's a collection of articles revolving around the topic _Is TDD dead?_:
- http://david.heinemeierhansson.com/2014/tdd-is-dead-long-live-testing.html
- http://rbcs-us.com/documents/Why-Most-Unit-Testing-is-Waste.pdf
- http://martinfowler.com/articles/is-tdd-dead/
- http://pythontesting.net/agile/is-tdd-dead/

**Pull request merged**